### PR TITLE
Reports Pagination

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Review by mmrudling — COMMENTED** [review]
vewwy nice

**Review by mmrudling — APPROVED** [review]

**Review by mmrudling — APPROVED** [review]
vewwy nice 2

File: AgentDiscoveries-Backend/src/main/java/org/softwire/training/utils/Pager.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            endPage -= (startPage - 1);
+            startPage = 1;
+        }
+
+        if (endPage > totalPages) {
+            endPage = totalPages;
+            if (endPage > 10) {
+                startPage = endPage - 9;
+            }
+        }
```
Comment (Kenneth-Siu) [inline AgentDiscoveries-Backend/src/main/java/org/softwire/training/utils/Pager.java:28]: Can we make some unit tests for this?

**Review by Kenneth-Siu — CHANGES_REQUESTED** [review]
Looks clean as always, just need to add some tests for your Pager now!

**Review by Kenneth-Siu — CHANGES_REQUESTED** [review]
The things we mentioned by talking

File: AgentDiscoveries-Backend/src/test/java/org/softwire/training/utils/PagerTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Test
+    void calculatesNumOfPages() {
+        int totalPages = pager.getTotalPages();
+
+        assertEquals(10, totalPages);
+    }
+
+    @Test
+    void calculatesStartPage() {
```
Comment (Kenneth-Siu) [inline AgentDiscoveries-Backend/src/test/java/org/softwire/training/utils/PagerTest.java:46]: I don't think this test actually tests the general case, we only test the edge case at the moment.

Can we also make the test name describe what the start page should be? (i.e. should be five before the current page)

Writing good tests is a fine art! Sorry if I sound a bit picky...

**Review by Kenneth-Siu — CHANGES_REQUESTED** [review]

File: AgentDiscoveries-Backend/src/test/java/org/softwire/training/utils/PagerTest.java [hunk 3]
```diff
@@ -56,6 +56,15 @@ void calculatesEndPage() {
         assertEquals(10, endPage);
     }
 
+    @Test
+    void startPageIsFivePagesBeforeCurrentPageAfterPageFive() {
+        pager = new Pager(200, 6, 10);
```
Comment (Kenneth-Siu) [inline AgentDiscoveries-Backend/src/test/java/org/softwire/training/utils/PagerTest.java]: FWIW, I'd pick 7 instead of 6, so that the start page isn't 1 (in theory someone could slip up, make it start six pages previous, and then this test would still pass because of your other edge condition)

I'm sure you're sick of listening to me now 😉 so once that's sorted, you can merge it straight in

**Review by Kenneth-Siu — APPROVED** [review]
Remember to make the final change and then you can just merge :)
